File: src/main/java/wraith/fabricaeexnihilo/json/basic/BlockJson.java
package wraith.fabricaeexnihilo.json.basic;

import com.google.gson.*;
import net.minecraft.block.Block;
import net.minecraft.util.Identifier;
import net.minecraft.util.registry.Registry;
import wraith.fabricaeexnihilo.json.BaseJson;
import wraith.fabricaeexnihilo.json.other.IdentifierJson;
import wraith.fabricaeexnihilo.util.RegistryUtils;

import java.lang.reflect.Type;

public final class BlockJson extends BaseJson<Block> {

    private BlockJson() {
    }

    public static final BlockJson INSTANCE = new BlockJson();

    @Override
    public Block deserialize(JsonElement json, Type typeOfT, JsonDeserializationContext context) throws JsonParseException {
        return Registry.BLOCK.get(context.<Identifier>deserialize(json, IdentifierJson.TYPE_TOKEN));
    }

    @Override
    public JsonElement serialize(Block src, Type typeOfSrc, JsonSerializationContext context) {
        return new JsonPrimitive(RegistryUtils.getId(src).toString());
    }

}

File: src/main/kotlin/exnihilofabrico/json/BlockJson.kt
package exnihilofabrico.json

import com.google.gson.*
import net.minecraft.block.Block
import net.minecraft.util.Identifier
import net.minecraft.util.registry.Registry
import java.lang.reflect.Type

object BlockJson: JsonDeserializer<Block>, JsonSerializer<Block> {
    override fun deserialize(json: JsonElement, typeOfT: Type, context: JsonDeserializationContext) =
        Registry.BLOCK[Identifier(json.asString)]

    override fun serialize(src: Block, typeOfSrc: Type, context: JsonSerializationContext) =
        JsonPrimitive(Registry.BLOCK.getId(src).toString())

}